File: LLD/meetingScheduler.java


/*
Meeting Room Scheduler
1) N meeting rooms with a given capacity
2) Book a meeting room for a given start time, end time and num_participants
    2.1 User can see list of all unoccupied meeting rooms for a given timeslot and capacity
3) Booking Strategy First-Come First-Serve basis
4) Send notification to all the participants of the meeting
5) History of all meetings scheduled
6) Check participants calendar for availability
7) Calendar to track bookings and meetings
8) Update meeting/Cancel meeting
    8.1 Update meeting participants - add participant (if cap crosses?)

Entities: Meeting, MeetingRoom, Participant, TimeSlot, Calendar, MeetingRoomController

Class Design Diagram:
---------------------
[MeetingScheduler] "1" o-- "1" [MeetingRoomController]
[MeetingScheduler] "1" o-- "1" [RoomSelectionStrategy]
[MeetingRoomController] "1" *-- "*" [MeetingRoom]
[MeetingRoom] "1" *-- "1" [Calendar]
[MeetingRoom] "1" *-- "*" [Meeting]
[Meeting] "1" *-- "*" [Participant]
[Meeting] "1" *-- "1" [TimeSlot]
[Participant] ..|> [Observer] : Implements
[FirstComeFirstServeStrategy] ..|> [RoomSelectionStrategy] : Implements

Class Details:
---------------------
1. MeetingScheduler
   - Role: Facade for scheduling logic.
   - Attributes: roomController, selectionStrategy.
   - Methods: scheduleMeeting().

2. MeetingRoomController
   - Role: Manages list of rooms and filters them.
   - Attributes: meetingRooms (List).
   - Methods: addMeetingRoom(), getAvailableMeetingRooms().

3. MeetingRoom
   - Role: Physical room entity.
   - Attributes: roomId, capacity, calendar, meetingsHistory.
   - Methods: isAvailable().

4. Meeting
   - Role: Event entity.
   - Attributes: id, title, participants, timeSlot.
   - Methods: notifyParticipants(), addParticipant().

5. Calendar
   - Role: Manages availability for a room or user.
   - Attributes: bookings (List<TimeSlot>).
   - Methods: isAvailable(), addTimeSlot().

6. Participant
   - Role: User attending meeting.
   - Methods: update() [Observer pattern].

7. RoomSelectionStrategy (Interface)
   - Role: Algorithm to pick best room from available list.
   - Methods: selectRoom().
*/

import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;

interface Observer {
    public void update(String message);
}
class Participant implements Observer {
    String name;
    String email;
    String mobile;
    Calendar calendar;

    public Participant(String name, String email, String mobile) {
        this.name = name;
        this.email = email;
        this.mobile = mobile;
        this.calendar = new Calendar();
    }

    @Override
    public void update(String message) {
        System.out.println("Email sent to " + this.email + ": " + message);
    }
}

class Meeting {
    String meetingId;
    String title;
    List<Participant> participants;
    TimeSlot timeSlot;

    public Meeting(String meetingId, String title, List<Participant> participants, TimeSlot timeSlot) {
        this.meetingId = meetingId;
        this.title = title;
        this.participants = participants;
        this.timeSlot = timeSlot;
    }

    public void addParticipant(Participant participant) {
        this.participants.add(participant);
    }

    public void removeParticipant(Participant participant) {
        this.participants.remove(participant);
    }
    public void notifyParticipants(String message) {
        for(Participant participant: participants) {
            participant.update(message);
        }
    }
}

class MeetingRoom {
    String roomId;
    List<Meeting> meetingsHistory;
    int capacity;
    Calendar calendar;
    Location location;

    public MeetingRoom(String roomId, Location location, int capacity) {
        this.roomId = roomId + "_" + location.floorId + "_" + location.buildingId;
        this.capacity = capacity;
        this.location = location;
        this.calendar = new Calendar();
        this.meetingsHistory = new ArrayList<>();
    }

    public boolean isAvailable(TimeSlot slot) {
        return this.calendar.isAvailable(slot);
    }
}

class TimeSlot {
    int startTime;
    int endTime;


    public TimeSlot(int startTime, int endTime) {
        this.startTime = startTime;
        this.endTime = endTime;
    }

    public boolean isOverlapping(TimeSlot other) {
        return this.startTime < other.endTime && this.endTime > other.startTime;
    }
}

class Location {
    int floorId;
    int buildingId;

    public Location(int floorId, int buildingId) {
        this.floorId = floorId;
        this.buildingId = buildingId;
    }
}

class Calendar {
    List<TimeSlot> bookings;

    public Calendar() {
        this.bookings = new ArrayList<>();
    }

    public void addTimeSlot(TimeSlot slot) {
        this.bookings.add(slot);
    }

    public void removeTimeSlot(TimeSlot slot) {
        this.bookings.remove(slot);
    }

    public boolean isAvailable(TimeSlot slot) {
        if(this.bookings.isEmpty()) {
            return true;
        }
        for(TimeSlot booking: this.bookings) {
            if(slot.isOverlapping(booking)) {
                return false;
            }
        }
        return true;
    }
}

interface RoomSelectionStrategy {
    public MeetingRoom selectRoom(List<MeetingRoom> meetingRooms);
}

class FirstComeFirstServeStrategy implements RoomSelectionStrategy {
    @Override
    public MeetingRoom selectRoom(List<MeetingRoom> availableRooms) {
        if(!availableRooms.isEmpty()) {
            return availableRooms.get(0);
        }
        return null;
    }
}
class MeetingRoomController {
    List<MeetingRoom> meetingRooms = new ArrayList<>();

    public void addMeetingRoom(MeetingRoom room) {
        this.meetingRooms.add(room);
        System.out.println("Add Rooms: " + Arrays.toString(this.meetingRooms.toArray()));
    }

    public List<MeetingRoom> getAvailableMeetingRooms(int startTime, int endTime, int numParticipants) {
        List<MeetingRoom> availableRooms = new ArrayList<>();
        for(MeetingRoom room: this.meetingRooms) {
            System.out.println("Room: " + room + " " + room.capacity + " " + numParticipants);
            if(room.capacity >= numParticipants && room.isAvailable(new TimeSlot(startTime, endTime))) {
                availableRooms.add(room);
            }
        }
        return availableRooms;
    }
}

class MeetingScheduler {
    RoomSelectionStrategy selectionStrategy;
    MeetingRoomController roomController;
    public MeetingScheduler(RoomSelectionStrategy selectionStrategy, MeetingRoomController roomController) {
        this.roomController = roomController;
        this.selectionStrategy = selectionStrategy;
    }

    public MeetingRoom scheduleMeeting(String meetingId, String title, List<Participant> participants, TimeSlot timeSlot) {
        List<MeetingRoom> availableRooms = roomController.getAvailableMeetingRooms(timeSlot.startTime, timeSlot.endTime, participants.size());
        System.out.println("Available Rooms: " + Arrays.toString(availableRooms.toArray()));
        MeetingRoom selectedRoom = selectionStrategy.selectRoom(availableRooms);

        if(selectedRoom != null) {
            Meeting newMeeting = new Meeting(meetingId, title, participants, timeSlot);
            selectedRoom.meetingsHistory.add(newMeeting);
            selectedRoom.calendar.addTimeSlot(timeSlot);
            newMeeting.notifyParticipants("Meeting scheduled in room " + selectedRoom.roomId + " from " + timeSlot.startTime + " - " + timeSlot.endTime);
            System.out.println("Meeting scheduled in room " + selectedRoom.roomId + " from " + timeSlot.startTime + " - " + timeSlot.endTime);
            return selectedRoom;
        } else {
            System.out.println("No room available for given time slot and capacity");
            return null;
        }
    }
}

class Main {
    public static void main(String[] args) {

        MeetingRoom room1 = new MeetingRoom("Room1", new Location(1,101), 3);
        MeetingRoom room2 = new MeetingRoom("Room2", new Location(1,101), 2);
        MeetingRoom room3 = new MeetingRoom("Room3", new Location(1,101), 5);

        MeetingRoomController controller = new MeetingRoomController();

        // Add meeting rooms
        controller.addMeetingRoom(room1);
        controller.addMeetingRoom(room2);
        controller.addMeetingRoom(room3);

        // Create participants
        Participant p1 = new Participant("Shreya", "abc@gmail.com", "+91234567890");
        Participant p2 = new Participant("Shravani", "xyz@gmail.com", "+91234567890");

        List<Participant> participants = new ArrayList<>();
        participants.add(p1);
        participants.add(p2);

        // Create a meeting
        MeetingScheduler scheduler = new MeetingScheduler(new FirstComeFirstServeStrategy(), controller);
        scheduler.scheduleMeeting("M1", "Archiving Discussion", participants, new TimeSlot(2, 3));
        scheduler.scheduleMeeting("M2", "Workflow Discussion", participants, new TimeSlot(2, 3));
    }
}

File: LLD/MeetingScheduler.py
# fmt: off
# ==============================================================================
#  MEETING SCHEDULER — ASCII CLASS DIAGRAM
# ==============================================================================
#
#  ┌──────────────────────────────────────────────────────────────────────────┐
#  │                        MEETING SCHEDULER                                 │
#  └──────────────────────────────────────────────────────────────────────────┘
#
#  ┌────────────────────────────────────┐
#  │         MeetingScheduler           │  ← Facade
#  ├────────────────────────────────────┤
#  │ + room_controller: MeetingRoomCtrl │
#  │ + strategy: RoomSelectionStrategy  │
#  ├────────────────────────────────────┤
#  │ + schedule(title, participants,    │
#  │            slot): Optional[Meeting]│
#  └──────────┬─────────────────────────┘
#             │ 1 uses                  │ 1 uses
#             ▼                        ▼
#  ┌──────────────────────────┐  ┌──────────────────────────────┐
#  │  MeetingRoomController   │  │    RoomSelectionStrategy     │
#  ├──────────────────────────┤  │        (ABC/Interface)       │
#  │ + rooms: Dict            │  ├──────────────────────────────┤
#  ├──────────────────────────┤  │ + select_room(rooms, count,  │
#  │ + add_room()             │  │    slot): Optional[Room]     │
#  │ + get_all_rooms()        │  └──────────────┬───────────────┘
#  └──────────┬───────────────┘                 │
#             │ 1..*                            ▼
#             ▼                  ┌──────────────────────────────┐
#  ┌──────────────────────────┐  │   FirstComeFirstServeStrategy│
#  │       MeetingRoom        │  ├──────────────────────────────┤
#  ├──────────────────────────┤  │ + select_room(): picks first │
#  │ + id: str                │  │   free room that fits count  │
#  │ + capacity: int          │  └──────────────────────────────┘
#  │ + calendar: Calendar     │
#  └──────────────────────────┘
#
#  ┌─────────────────────────────────────────┐
#  │              Meeting                    │
#  ├─────────────────────────────────────────┤
#  │ + id: str                               │
#  │ + title: str                            │
#  │ + participants: List[Participant]       │
#  │ + room: MeetingRoom                     │
#  │ + slot: TimeSlot                        │
#  └─────────────────────────────────────────┘
#
#  ┌──────────────────────┐    ┌─────────────────────────────────────┐
#  │     Participant      │    │              Calendar               │
#  ├──────────────────────┤    ├─────────────────────────────────────┤
#  │ + id: str            │    │ + bookings: List[TimeSlot]          │
#  │ + name: str          │    │ - _lock: Lock                       │
#  │ + email: str         │    ├─────────────────────────────────────┤
#  │ + calendar: Calendar │    │ + is_available(slot): bool          │
#  ├──────────────────────┤    │ + add_booking(slot)                 │
#  │ + notify(message)    │    │ + remove_booking(slot)              │
#  └──────────────────────┘    └─────────────────────────────────────┘
#
#  ┌──────────────────────────────┐
#  │          TimeSlot            │
#  ├──────────────────────────────┤
#  │ + start_time: datetime       │
#  │ + end_time: datetime         │
#  ├──────────────────────────────┤
#  │ + overlaps(other): bool      │
#  └──────────────────────────────┘
#
#  RELATIONSHIPS:
#  MeetingScheduler ──1──> MeetingRoomController   (aggregates rooms)
#  MeetingScheduler ──1──> RoomSelectionStrategy   (pluggable strategy)
#  MeetingRoomController ──*──> MeetingRoom        (manages rooms)
#  Meeting  ──1──> MeetingRoom                     (booked in a room)
#  Meeting  ──*──> Participant                     (attendees)
#  Participant ──1──> Calendar                     (personal schedule)
#  MeetingRoom ──1──> Calendar                     (room schedule)
#  TimeSlot.overlaps() detects scheduling conflicts
#  Observer: Participant.notify() called for every scheduled meeting
# ==============================================================================
# fmt: on
import threading
import uuid
from abc import ABC, abstractmethod
from typing import List, Dict, Optional
from datetime import datetime

"""
==============================================================================================
MEETING SCHEDULER LOW LEVEL DESIGN (INTERVIEW OPTIMIZED)
==============================================================================================

Key Features:
1. Resource Management: Meeting rooms with varying capacities.
2. Participant Management: Track availability via personal calendars.
3. Scheduling Logic: Finds available rooms and checks participant conflicts.
4. Strategy Pattern: Pluggable room selection logic (e.g., FCFS, best-fit).
5. Observer Pattern: Notifies participants of scheduled meetings.
6. Thread-safety: Synchronized access to room availability and scheduling.

Design Patterns:
1. Facade: MeetingScheduler (Central Controller).
2. Strategy: RoomSelectionStrategy.
3. Observer: Participant notification.

Class Design Diagram:
---------------------
[MeetingScheduler] "1" o-- "1" [MeetingRoomController]
[MeetingScheduler] "1" o-- "1" [RoomSelectionStrategy]
[MeetingRoomController] "1" *-- "*" [MeetingRoom]
[MeetingRoom] "1" *-- "1" [Calendar]
[Meeting] "1" *-- "*" [Participant]
[Meeting] "1" *-- "1" [TimeSlot]
[Participant] ..|> [Observer] : Implements
[FirstComeFirstServeStrategy] ..|> [RoomSelectionStrategy]

Class Details:
---------------------
1. MeetingScheduler (Facade)
   - Role: Facade for scheduling logic.
   - Attributes: roomController, selectionStrategy.
   - Methods: schedule().

2. MeetingRoomController
   - Role: Manages list of rooms and filters them.
   - Methods: addMeetingRoom(), getAllRooms().

3. MeetingRoom
   - Role: Physical room entity.
   - Attributes: roomId, capacity, calendar.

4. Meeting
   - Role: Event entity.
   - Attributes: id, title, participants, timeSlot, room.

5. Calendar
   - Role: Manages availability for a room or user.
   - Attributes: bookings (List<TimeSlot>).
   - Methods: isAvailable(), addBooking().

6. Participant
   - Role: User attending meeting. Receives notifications.
"""

# ==========================================
# Domain Models
# ==========================================

class TimeSlot:
    """Represents a duration of time."""
    def __init__(self, start_time: datetime, end_time: datetime):
        if start_time >= end_time:
            raise ValueError("Start time must be before end time.")
        self.start_time = start_time
        self.end_time = end_time

    def overlaps(self, other: 'TimeSlot') -> bool:
        return self.start_time < other.end_time and self.end_time > other.start_time

    def __repr__(self):
        return f"{self.start_time.strftime('%H:%M')} - {self.end_time.strftime('%H:%M')}"

class Calendar:
    """Manages a schedule of time slots."""
    def __init__(self):
        self.bookings: List[TimeSlot] = []
        self._lock = threading.Lock()

    def is_available(self, slot: TimeSlot) -> bool:
        with self._lock:
            return not any(slot.overlaps(b) for b in self.bookings)

    def add_booking(self, slot: TimeSlot):
        with self._lock:
            self.bookings.append(slot)

    def remove_booking(self, slot: TimeSlot):
        with self._lock:
            if slot in self.bookings:
                self.bookings.remove(slot)

class Participant:
    """A person attending meetings."""
    def __init__(self, name: str, email: str):
        self.id = str(uuid.uuid4())
        self.name = name
        self.email = email
        self.calendar = Calendar()

    def notify(self, message: str):
        print(f"INFO: Notification -> {self.name} ({self.email}): {message}")

class MeetingRoom:
    """Physical meeting room entity."""
    def __init__(self, room_id: str, capacity: int):
        self.id = room_id
        self.capacity = capacity
        self.calendar = Calendar()

    def __repr__(self):
        return f"Room({self.id}, Cap: {self.capacity})"

class Meeting:
    """An event scheduled in the system."""
    def __init__(self, title: str, participants: List[Participant], room: MeetingRoom, slot: TimeSlot):
        self.id = str(uuid.uuid4())
        self.title = title
        self.participants = participants
        self.room = room
        self.slot = slot

# ==========================================
# Scheduling Strategies
# ==========================================

class RoomSelectionStrategy(ABC):
    @abstractmethod
    def select_room(self, rooms: List[MeetingRoom], count: int, slot: TimeSlot) -> Optional[MeetingRoom]:
        pass

class FirstComeFirstServeStrategy(RoomSelectionStrategy):
    """Picks the first available room that fits the participant count."""
    def select_room(self, rooms: List[MeetingRoom], count: int, slot: TimeSlot) -> Optional[MeetingRoom]:
        for room in rooms:
            if room.capacity >= count and room.calendar.is_available(slot):
                return room
        return None

# ==========================================
# Meeting Room Controller
# ==========================================

class MeetingRoomController:
    """Manages the inventory of meeting rooms."""
    def __init__(self):
        self.rooms: Dict[str, MeetingRoom] = {}

    def add_room(self, room: MeetingRoom):
        self.rooms[room.id] = room
        print(f"INFO: Added meeting room: {room.id}")

    def get_all_rooms(self) -> List[MeetingRoom]:
        return list(self.rooms.values())

# ==========================================
# Meeting Scheduler (Facade)
# ==========================================

class MeetingScheduler:
    """Facade for meeting room scheduling."""
    def __init__(self):
        self.room_controller = MeetingRoomController()
        self.strategy: RoomSelectionStrategy = FirstComeFirstServeStrategy()
        print("INFO: MeetingScheduler initialized.")

    def schedule(self, title: str, participants: List[Participant], slot: TimeSlot) -> Optional[Meeting]:
        """Schedules a meeting if a room and all participants are available."""
        # 1. Check participant availability
        for p in participants:
            if not p.calendar.is_available(slot):
                print(f"WARNING: {p.name} is busy during {slot}")
                return None

        # 2. Find an available room
        room = self.strategy.select_room(self.room_controller.get_all_rooms(), len(participants), slot)
        if not room:
            print(f"WARNING: No available room for {len(participants)} participants at {slot}")
            return None

        # 3. Book room and participants
        room.calendar.add_booking(slot)
        for p in participants:
            p.calendar.add_booking(slot)

        # 4. Create and notify
        meeting = Meeting(title, participants, room, slot)
        msg = f"Meeting '{title}' scheduled in {room.id} at {slot}."
        for p in participants:
            p.notify(msg)
        print(f"INFO: {msg}")
        return meeting

# ==========================================
# Main Execution / Demo
# ==========================================

if __name__ == "__main__":
    print("--- Starting Meeting Scheduler Demo ---")

    scheduler = MeetingScheduler()
    scheduler.room_controller.add_room(MeetingRoom("Conf-Room-A", 10))
    scheduler.room_controller.add_room(MeetingRoom("Huddle-Box-1", 3))

    p1 = Participant("Alice", "alice@corp.com")
    p2 = Participant("Bob", "bob@corp.com")
    p3 = Participant("Charlie", "charlie@corp.com")

    print("\n[Scenario 1] Alice and Bob want a 1-hour sync at 2 PM")
    slot1 = TimeSlot(datetime(2025, 1, 1, 14, 0), datetime(2025, 1, 1, 15, 0))
    scheduler.schedule("Product Sync", [p1, p2], slot1)

    print("\n[Scenario 2] Charlie wants to meet Bob at 2:30 PM (Conflict for Bob)")
    slot2 = TimeSlot(datetime(2025, 1, 1, 14, 30), datetime(2025, 1, 1, 15, 30))
    scheduler.schedule("Tech Review", [p2, p3], slot2)

    print("\n[Scenario 3] Charlie and Alice meet at 4 PM")
    slot3 = TimeSlot(datetime(2025, 1, 1, 16, 0), datetime(2025, 1, 1, 17, 0))
    scheduler.schedule("Design Brainstorm", [p1, p3], slot3)
